lec15-git-projects.Rmd

```diff
@@ -477,6 +478,233 @@ The 'GitHub flow':
+members of your group.
+
+4. Once your changes have been approved, merge the PR. It's good practice to
+   have group members merge your PRs instead of doing it yourself. 
+
+Although this process may seem a bit laborious, using this method (also known
+as the 'GitHub flow') minimizes chances of error and ensures that all code is
+reviewed by at least one other person. Understanding how and why this process
+works is key to collaborative work in software development and the like, and is
+used by all sorts of open source projects on GitHub ([including `dplyr`
+itself](https://github.com/tidyverse/dplyr)!)
+
+#### Creating a new branch
+
+In a given repo, there is a 'Branch: master' button just above the list of
+files. Clicking on this will bring up a dropdown menu containing a list of
+branches (which currently just contains `master`) and a text field.  A new
+branch can be created by typing in the text field. Start by making a new branch
+called `initial-edit` using said text field. GitHub will then refresh, and the
+button will now say 'Branch: initial-edit' instead.
+
+![_Pictured: the branch dropdown menu_](image/git_lesson/branch_dropdown.png)
+
+The dropdown can now be used to switch back to `master`, but let's stay in
+`initial-edit` for now. This is the place to make new changes -- either by
+editing existing files or creating new ones. Let's create a simple Rmd file
+with the 'Create new file' button on the right side of the repo page.
+
+````
+---
+title: "Test R Notebook"
+author: "Your Name"
+output: pdf_document
+---
+
+Hello! This is my R notebook.
+
+```{r}`r ''`
+print('Hello, GitHub!')
+```
+
+````
+
+![_Pictured: editing a file on GitHub_](image/git_lesson/sample_rmd.png)
+
+Once done, scroll to the bottom of the page to save your changes. In GitHub
+lingo, a saved change is called a **commit** -- this can be thought of more as
+a 'checkpoint'.  GitHub _requires_ that all commits have an associated commit
+message describing the change.  This may seem a bit draconian, but it ensures
+that all changes have been documented, which is something that may save you and
+your group members a lot of trouble down the line!
+
+While there is a default commit message GitHub will fill in, it's best to add a
+custom one. In this case, something like 'initial Rmd file creation' will do.
+Click on 'Commit new file' to complete the process. Note that the radio button
+underneath specifically says 'Commit directly to the `initial-edit` branch'.
+This means that your edit will only exist in `initial-edit` for now -- `master`
+remains unchanged. 
+
+#### Pull requests
+
+After a commit has been made, GitHub takes you back to the main repo page, but
+now with an attention-grabbing yellow prompt about these new edits. GitHub has
+noticed that there exists a new branch with edits and that these edits differ
+from what's in `master`. Over on the right side of the prompt, GitHub presents
+the option of creating a pull request.
```

Comment:
May be worth mentioning that you can add multiple commits to the same branch, which may involve multiple files, before submitting a PR.